WordPress/openverse
# Add proportional staging index dag

## Fixes



Fixes #3488 by @stacimc 

## Description



This PR adds a new `create_proportional_by_source_staging_index` DAG as described in the IP [here](https://docs.openverse.org/projects/proposals/search_relevancy_sandbox/20230530-implementation_plan_staging_elasticsearch_reindex_dags.html#create-proportional-by-provider-media-type-staging-index-dag):
<img width="1398" alt="Screenshot 2024-02-07 at 12 53 35 PM" src="https://github.com/WordPress/openverse/assets/63313398/e75df133-cc4d-454e-a8d3-e11e4134d5da">

It can be used to create a new index in staging which is a subset of a staging index, but maintains the same proportion of records per source.

Note: reindexing is not parallelized with slicing, as it is for `create_new_production_es_index` and other DAGs (see comment thread for explanation). However it is parallelized in the sense that the individual reindexing tasks for each source can run in parallel. Both of these are acceptable because the DAG only touches staging indices, not production.


## Testing Instructions


Tests assume you're starting from fresh local testing data, ie after running `just api/init`.

* Run the `create_proportional_by_source_staging_index` DAG locally with default conf options. Then verify in elasticvue that you have a new es index called `audio-50-percent-proportional-20240214t181238` (the final suffix is a timestamp and so will differ), with the alias `audio-subset-by-source`.
  * By default it will be based on the `audio-filtered` index and should have 50% of the total documents, with the same source proportions. Since the filtered index has 4912 documents, the new index should have 2456.
  * Inspect the logs for `get_staging_source_counts` to see the counts for the source index, and then the `wait_for_reindex` tasks to see how many were reindexed for each source in the destination index. The proportions should remain the same. For example `wikimedia_audio` has 3992 records in the source index, and 1996 in the new one.
  * Note that the `remove_existing_alias` step was skipped in this dagrun
* Run the DAG again, but this time change the `percentage_of_prod` to 0.25 and the `source_index` to `audio`.
  * You should get a new index named something like `audio-25-percent-proportional-20240214t182518`
  * No steps should be skipped this time. The `audio-subset-by-source` alias should have been removed from the previous index and is now applied to this one.
  * The index should have 1250 documents (1/4 of 5000 from the main audio index)
* Run the DAG with media type `image` and ensure this also works. You should get a new index with `image-subset-by-source` alias, a name like `image-50-percent-proportional-20240214t181238`, and 2328 documents (half of the 4656 that are in the filtered image index). Check the proportions are correct.
* Run the `staging_elasticsearch_cluster_healthcheck` and `create_new_staging_es_index` DAGs to ensure that refactoring out shared elasticsearch utilities did not break them.

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
